module/VuFind/src/VuFind/Auth/OpenIDConnect.php

```diff
@@ -0,0 +1,529 @@
+        return $user;
+    }
+
+    /**
+     * Get the URL to establish a session (needed when the internal VuFind login form is inadequate). Returns false
+     * when no session initiator is needed.
+     *
+     * @param string $target Full URL where external authentication method should send user after login (some drivers
+     * may override this).
+     *
+     * @return bool|string
+     * @throws AuthException
+     */
+    public function getSessionInitiator($target): bool|string
+    {
+        // Adding the auth_method setting makes it possible to handle logins when
+        // using an auth method that proxies others (e.g. ChoiceAuth)
+        $targetUri = $target . (str_contains($target, '?') ? '&' : '?') . 'auth_method=oidc';
+        if (empty($this->session->oidcLastUri) && !empty($target)) {
+            $this->session->oidcLastUri = $targetUri;
+        }
+        $params = [
+            'response_type' => 'code',
+            'redirect_uri' => $targetUri,
+            'client_id' => $this->getConfig('client_id'),
+            'nonce' => $this->session->oidc_nonce,
+            'state' => $this->session->oidc_state,
+            'scope' => 'openid profile email',
+        ];
+        return $this->getProvider()->authorization_endpoint . '?' . http_build_query($params);
+    }
+
+    /**
+     * Obtain an access token from a code.
+     *
+     * @param string $code Code to look up.
+     *
+     * @return object
+     * @throws AuthException
+     */
+    protected function getRequestToken(string $code): object
+    {
+        if (isset($this->requestToken)) {
+            return $this->requestToken;
+        }
+        $provider = $this->getProvider();
+        $url = $provider->token_endpoint;
+        $params = [
+           'grant_type' => 'authorization_code',
+           'code' => $code,
+           'redirect_uri' => $this->session->oidcLastUri,
+           'client_id' => $this->getConfig('client_id'),
+           'client_secret' => $this->getConfig('client_secret'),
+        ];
+        $authMethods = $provider->token_endpoint_auth_methods_supported ?? null;
+        $headers = [];
+        if (in_array('client_secret_basic', $authMethods) || null === $authMethods) {
+            $headers = [
+                'Authorization: Basic ' . base64_encode(
+                    urlencode($this->getConfig('client_id')) . ':'
+                    . urlencode($this->getConfig('client_secret'))
+                ),
+            ];
+            unset($params['client_secret']);
+        }
+
+        $response = $this->httpService->post(
+            $url,
+            http_build_query($params),
+            'application/x-www-form-urlencoded',
+            null,
+            $headers
+        );
+        $json = json_decode($response->getBody());
+        if (isset($json->error)) {
+            throw new AuthException('authentication_error' . ': ' . $json->error_description ?? $json->error);
```

Comment:
I believe that `$json->error_description ?? $json->error` needs to be enclosed in parenthesis. Otherwise `.` takes precedence.

Reply:
Fixed